Towny/src/main/java/com/palmergames/bukkit/towny/utils/SpawnUtil.java

```diff
@@ -440,19 +445,88 @@ else if (town != null && town.hasSpawn())
 							return player.getWorld().getSpawnLocation();
 					});
 				} else if (town != null && town.hasSpawn())
-					yield CompletableFuture.completedFuture(town.getSpawnOrNull());
+					yield CompletableFuture.completedFuture(getSafeLocation(town.getSpawnOrNull()));
 				else
 					yield CompletableFuture.completedFuture(player.getWorld().getSpawnLocation());
 			case TOWN:
 				if (outpost)
-					yield CompletableFuture.completedFuture(getOutpostSpawnLocation(town, split));
+					yield CompletableFuture.completedFuture(getSafeLocation(getOutpostSpawnLocation(town, split)));
 				else
-					yield CompletableFuture.completedFuture(town.getSpawn());
+					yield CompletableFuture.completedFuture(getSafeLocation(town.getSpawn()));
 			case NATION:
-				yield CompletableFuture.completedFuture(nation.getSpawn());
+				yield CompletableFuture.completedFuture(getSafeLocation(nation.getSpawn()));
 		};
 	}
 
+	/**
+	 * Tries to find a safe location nearby to teleport the player to
+	 * if safety teleport is enabled, otherwise does nothing.
+	 * 
+	 * @param location Starting location
+	 * @return A safe location nearby, same location if already safe
+	 */
+	private static Location getSafeLocation(Location location) {
+		//if safety teleport isn't enabled do nothing
+		if (!TownySettings.getSafetyTeleport()) {
+			return location;
+		}
+		
+		if (isSafeLocation(location)) {
+			return location;
+		}
+
+		Location up = location.clone();
+		Location down = location.clone();
+		
+		// look for 20 blocks up and down for a safe location, 
+		// if you can't find it fail the teleport
+		// maybe add a translation key and add the player 
+		// as a parameter to print him an error message
+		for(int i = 0; i < 20; i++) {
+			up = up.add(0,1,0);
+			down = down.subtract(0, 1, 0);
+			
+			if (isSafeLocation(up)) {
+				return up;
+			}
+			
+			if (isSafeLocation(down)) {
+				return down;
+			}
+		}
+		
+		return null;

```

Comment:
I don't know if returning null is a good idea, probably better to return the original location.

The downside of that is someone wanting a killbox just makes a killbox that is 20 blocks up and down from the spawn and they still have a killbox.

If you're going to return null you should use the @nullable annotation and mention it is possible in the javadoc `@return` and then see if the code base handles a null spawn well.

Reply:
The main idea was to return null and display an error message that says to the player that he can't be teleported to that location, do you prefer the player to be teleported to an unsafe location and skip the error message part?

Reply:
in this:

```java
		getSpawnLoc(player, town, nation, spawnType, outpost, split).thenAccept(spawnLoc -> {
			if (!spawnLoc.isWorldLoaded()) {
				TownyMessaging.sendErrorMsg(player, Translatable.of("msg_err_world_not_loaded"));
				return;
			}
```
I'm fairly sure that this is where a null Location being returned would get caught.

Looking at the existing getSpawnLoc() method I can see a number of places that we aren't making sure we're not returning a null location. Your method returning null here is probably alright for that reason.

Reply:
I don't think so, when i tested the and went outside he range and tried to teleport there was no error message at all, if it was catched by this if it would have printed me an error

Reply:
Don't forget that CompletableFutures swallow exceptions by default, nothing would've been printed. In this case it is safe however because a thenAccept doesn't run if the result was null.